src/utils/compileJsx.ts

```diff
@@ -1,19 +1,51 @@
 import { transform } from 'sucrase';
+import { transform as transformWithBabel } from '@babel/standalone';
+import memoizeOne from 'memoize-one';
 
-export const openFragmentTag = '<React.Fragment>';
-export const closeFragmentTag = '</React.Fragment>';
+export const ReactFragmentPragma = 'R_F';
+export const ReactCreateElementPragma = 'R_cE';
 
-export const compileJsx = (code: string) =>
-  transform(`${openFragmentTag}${code.trim() || ''}${closeFragmentTag}`, {
-    transforms: ['jsx'],
-    production: true,
-  }).code;
+export const openFragmentTag = '<>';
+export const closeFragmentTag = '</>';
 
-export const validateCode = (code: string) => {
+const wrapInFragment = (code: string) =>
+  `${openFragmentTag}${code}${closeFragmentTag}`;
+
+// This one sometimes throws errors with less useful information, but is fast
+export const compileJsx = memoizeOne(
+  (code: string) =>
+    transform(wrapInFragment(code.trim()), {
+      transforms: ['jsx'],
+      jsxPragma: ReactCreateElementPragma,
+      jsxFragmentPragma: ReactFragmentPragma,
+      production: true,
+    }).code
+);
+
+// This one throws errors with line numbers. Useful for validation
+const compileJsxWithBabel = memoizeOne((code: string) =>
+  transformWithBabel(wrapInFragment(code), {
+    presets: [
+      [
+        'react',
+        {
+          development: false,
+          pure: false,
+          pragma: ReactCreateElementPragma,
+          pragmaFrag: ReactFragmentPragma,
+          runtime: 'classic',
+          useBuiltIns: true,
+        },
+      ],
+    ],
+  })
+);
+
+export const validateCode = (code: string): true | Error => {
```

Comment:
Is there a specific reason for returning this mix of types? The only usage of the error as I see it is up in the code editor, and this was refactored away from a try/catch.

I think this may just be a refactoring relic from compiling with sucrase.

[code editor]: https://github.com/seek-oss/playroom/pull/304/files#diff-daeb499fe7d4158c6642e7314093f1d6623e137ee5a91c50932a113692eb0677R47

Reply:
`validateCode` calls `compileJsxWithBabel`, not `compileJsx`. I didn't want to expose `compileJsxWithBabel`, so I merged the return types.